### ISSUE TITLE
nixos: Decide and document scope of configuration

### ISSUE THREAD

**Issue by cyounkins** [NixOS/nixpkgs#51630]
To what degree should configuration be handled in nix?

We have a number of related issues:
- Purity in /etc (#3192 #22005)
- Only a subset of options are exposed in nixpkgs (Compare https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/ssh/sshd.nix and https://linux.die.net/man/5/sshd_config), many GitHub issues created for exposing options (#30475 #33282)
- Unclear handling of defaults in options (https://github.com/NixOS/nixpkgs/pull/50849#pullrequestreview-176969593)
- Configurations can depend on `/nix/store` paths (eg systemd unit path to binary, lighthttpd config wants path to perl)

Some questions:
- Should we describe configuration options in nixpkgs? All or a subset?
- Do we generate configuration in Nix then write it out, or attempt to patch a copied config?
- Can we reasonably make `/etc` readonly?
- What parts of a system should be mutable? Users? Services? Containers? Passwords? Network?

Some opinions:
- Mirroring upstream configuration options is too much maintenance burden, too hard to get right
- Allow arbitrary configuration options, validating/type checking where possible
- Implicitly use upstream defaults
- Services should "do the right thing" by default. Log to systemd journal, put mutable state in correct directories.
- It should be hard to get into a broken configuration (like bad network config), but it should be easy to get out of one.

Please post your thoughts and ideas!

Edit: Removed questions of bootstrapping

**Comment by endgame** [comment]
I know it's a mountain of effort, but what I really want to do is write `configuration.nix`, and have it configure everything system-level on that machine. Readonly `/etc` would be fantastic. Hardcoded store paths do not bother me. Not having to learn the configuration syntax for each and every thing is surprisingly nice.

That said, a useful meta-option might be to just override entire config files, for when you can't do something through nixpkgs _yet_, but you need to do it Right Now.

I wonder: how much work would it be to start automating the generation of options from specifications (like we do with haskell packages, and probably other packaging systems too).

It's not obvious to me whether a hypothetical bootstrap env should use old or new config. I can imagine situations where either makes sense, depending on the order of applying changes. This doesn't have a straightforward solution that i can see.

**Comment by danbst** [comment]
@endgame
> I wonder: how much work would it be to start automating the generation of options from specifications (like we do with haskell packages, and probably other packaging systems too).

probably best example of what happens when NixOS module options are generated is [nixsap](https://github.com/zalora/nixsap).

For example, postgresql options were translated to NixOS: https://github.com/zalora/nixsap/blob/master/modules/apps/postgresql/server.nix

Does it look like what you've imagined?

**Comment by endgame** [comment]
It looks like a pretty cool project, but where are the options coming from? Is there a manifest somewhere in that repo which is used to generate that `.../postgresql/server.nix` you mentioned?

**Comment by danbst** [comment]
Looks like I lied and these options are not autogenerated. cc @ip1981 on the generation method.

> I'm okay with things like .extraConfig options to append arbitrary text to a config file

It doesn't work in some scenarios.
1. nginx. One can't simply override/extend existing virtual host by appending stuff to config. That's why we have a tree of configuration options for it in NixOS. For most nested stuff dumb `.extraConfig` doesn't work.

Another example is with postgres:
```
    services.postgresql.extraConfig = ''
        listen_addresses = '127.0.0.1'
        listen_addresses = '127.0.0.2'
    '';
```
The latter line will override former, but I wanted merge behavior.
2. In some (rare) cases, it is forbidden to duplicate config entries, so even override behavior is absent

So, depending on config parsing of a specific service, we have to choose different strategies of option naming. `.extraConfig` is somtimes not sufficient.

**Comment by endgame** [comment]
Agreed that it doesn't solve every problem, and that having well-structured options is superior, but it does give you an escape hatch in a number of situations.

**Comment by danbst** [comment]
@cyounkins 

> Configurations can depend on /nix/store paths (eg systemd unit path to binary, lighthttpd config wants path to perl

There is nothing bad here, right? 

> Should we describe configuration options in nixpkgs? All or a subset?

My rules of thumb:
- if you want this configuration option to be discoverable by NixOS newcomers, expose it as NixOS option and write good documentation on how can it helps life
- if it is impossible to define in `.extraConfig` style, but it is of great importance, expose as an option
- otherwise provide `.configFile` and `.extraConfig`
- one more desire is to expose functions used to generate config. So that it would be possible to do https://github.com/NixOS/nixpkgs/pull/22209 entirely in `configuration.nix`. As of now it is extremely hard to do such overrides, have to fork module.

> Do we generate configuration in Nix then write it out, or attempt to patch a copied config?

Generate. Patching isn't much composable (but there can be hacks like
```
configPatch = mkOption {
  type = types.str; # note, not type.lines, but str, which forbids any merges
  example = ''
    sed 's/xxx/yyy/g' "$1"
  '';
  description = "run this patch over default configuration";
};
```
)

**Comment by danbst** [comment]
Related https://github.com/NixOS/rfcs/pull/42. Some comments discuss topics from OP-post

**Comment by Infinisil** [comment]
Thanks for pointing this out @danbst, the RFC is very relevant to this issue, it addresses multiple concerns mentioned here.

A structural `config` option should remove the need for stringly-typed `extraConfig` options. A structural version allows merging of multiple assigments correctly, introspection into its value, a flexible assignment of defaults and not needing hardcoded defaults. It also makes sure all upstream settings can be assigned, without having to maintain all of them in NixOS.

And with such an option, it then makes a lot of sense to limit the amount of additional NixOS options for specific settings to the ones that most people will be using, therefore decreasing the maintenance burden and module size.

I am currently in the process of rewriting a big part of the RFC in order to address the comments on it.

**Comment by cyounkins** [comment]
This is still important and I don't think it's really been resolved.

**Comment by davidak** [comment]
This is still important.

**Comment by buckley310** [comment]
rm stale

**Comment by dark-ether** [comment]
in my opinion having all configuration accessible through nix should be our goal. however that seems like a massive undertaking and even worse partial configuration through nix and normal configuration is probably inferior to configuring only through the normal method. so we probably should mark the ones we want "nixfied" and then create a module when enough interest arises
